src/D2L.CodeStyle.Analyzers/Async/Generator/FileCollector.cs

```diff
@@ -0,0 +1,304 @@
+﻿using System.Collections.Immutable;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace D2L.CodeStyle.Analyzers.Async.Generator;
+
+internal partial class SyncGenerator {
+	/// <summary>
+	/// Creates a source file containing generated methods and the necessary
+	/// namespaces, using directives, and type declarations for them to live in
+	/// (but only the necessary ones).
+	/// </summary>
+	internal sealed class FileCollector {
+		private static readonly SyntaxToken NoneToken
+			= SyntaxFactory.Token( SyntaxKind.None );
+
+		private static readonly SyntaxTriviaList EmptyTrivia
+			= SyntaxFactory.TriviaList();
+
+		private readonly CompilationUnitSyntax m_root;
+
+		// This is a mutable dictionary because we will remove from it as we go
+		// and check that it is empty at the end, just to be safe.
+		private readonly Dictionary<SyntaxNode, IEnumerable<string>> m_methods;
+
+		/// <summary>
+		/// Wraps the sort of syntax that we will append as part of a preamble, to
+		/// defer calls to ToFullString()
+		/// </summary>
+		private readonly struct PieceOfSyntax {
+			public object Data { get; init; }
+
+			public static implicit operator PieceOfSyntax( SyntaxNode node )
+				=> new() { Data = node };
+
+			public static implicit operator PieceOfSyntax( SyntaxToken token )
+				=> new() { Data = token };
+
+			public static implicit operator PieceOfSyntax( SyntaxList<UsingDirectiveSyntax> list )
+				=> new() { Data = list };
+
+			public override string ToString() => Data switch {
+				SyntaxNode n => n.ToFullString(),
+				SyntaxToken t => t.ToFullString(),
+				SyntaxList<UsingDirectiveSyntax> nl => nl.ToFullString(),
+				_ => throw new BugException( "invalid piece of syntax " + Data.GetType().Name )
+			};
+		}
+
+		// When we recurse into a namespace or type, we're not sure if the
+		// preamble ("namespace foo {", "\t\tpartial class Bar<T> {", etc.) is
+		// necessary, because maybe there are no generated methods.
+		// Using a LinkedList<string> as a double-ended queue.
+		private readonly LinkedList<PieceOfSyntax[]> m_preambles = new();
+
+		private readonly StringBuilder m_out = new();
+
+		private FileCollector(
+			CompilationUnitSyntax root,
+			Dictionary<SyntaxNode, IEnumerable<string>> methods
+		) {
+			m_root = root;
+			m_methods = methods;
+		}
+
+		public static FileCollector Create(
+			CompilationUnitSyntax root,
+			ImmutableArray<(SyntaxNode Parent, string Source)> methods

```

Comment:
Why take `string` for the method source here instead of something like `SyntaxNode`/`SyntaxTree`?

Reply:
In the full pipeline (not evident in this PR unfortunately) the methods are generated in parallel (with some caching) and collected into these files. That collection step causes some synchronization, and it seems to work better to split it up this way. Ultimately we need to get a `string` per file out (as that's what the generator APIs take).

Imagine you're in a file, making edits. Every edit will cause this code to run again, concatenating possibly cached generated sync methods. We need to do this concatenation on every edit, unfortunately, because edits outside of the method can impact the generated file (e.g. you change the usings, or nest/unnest namespaces types etc., basically the stuff we're I'm calling "preamble" isn't cachable within a file easily.)